src/main/java/org/apache/lucene/BoundedLinearFeatureQuery.java

```diff
@@ -0,0 +1,235 @@
+            return false;
+        }
+        BoundedLinearFeatureQuery that = (BoundedLinearFeatureQuery) obj;
+        return Objects.equals(fieldName, that.fieldName)
+            && Objects.equals(featureName, that.featureName)
+            && Objects.equals(scoreUpperBound, that.scoreUpperBound);
+    }
+
+    @Override
+    public int hashCode() {
+        int h = getClass().hashCode();
+        h = 31 * h + fieldName.hashCode();
+        h = 31 * h + featureName.hashCode();
+        h = 31 * h + scoreUpperBound.hashCode();
+        return h;
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        if (!scoreMode.needsScores()) {
+            // We don't need scores (e.g. for faceting), and since features are stored as terms,
+            // allow TermQuery to optimize in this case
+            TermQuery tq = new TermQuery(new Term(fieldName, featureName));
+            return searcher.rewrite(tq).createWeight(searcher, scoreMode, boost);
+        }
+
+        return new Weight(this) {
+
+            @Override
+            public boolean isCacheable(LeafReaderContext ctx) {
+                return false;
+            }
+
+            @Override
+            public Explanation explain(LeafReaderContext context, int doc) throws IOException {
+                String desc = "weight(" + getQuery() + " in " + doc + ") [\" BoundedLinearFeatureQuery \"]";
+
+                Terms terms = context.reader().terms(fieldName);
+                if (terms == null) {
+                    return Explanation.noMatch(desc + ". Field " + fieldName + " doesn't exist.");
+                }
+                TermsEnum termsEnum = terms.iterator();
+                if (termsEnum.seekExact(new BytesRef(featureName)) == false) {
+                    return Explanation.noMatch(desc + ". Feature " + featureName + " doesn't exist.");
+                }
+
+                PostingsEnum postings = termsEnum.postings(null, PostingsEnum.FREQS);
+                if (postings.advance(doc) != doc) {
+                    return Explanation.noMatch(desc + ". Feature " + featureName + " isn't set.");
+                }
+
+                int freq = postings.freq();
+                float featureValue = decodeFeatureValue(freq);
+                float score = boost * featureValue;
+                return Explanation.match(
+                    score,
+                    "Linear function on the " + fieldName + " field for the " + featureName + " feature, computed as w * S from:",
+                    Explanation.match(boost, "w, weight of this function"),
+                    Explanation.match(featureValue, "S, feature value")
+                );
+            }
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                Terms terms = Terms.getTerms(context.reader(), fieldName);
```

Comment:
what will happen in the case when the field with `feildName` is not present

Reply:
When creating this new object, the fieldName is required non null.